packages/dialog/src/vaadin-dialog.js

```diff
@@ -53,18 +97,188 @@ export class DialogOverlay extends OverlayElement {
+    this._titleId = `${this.constructor.is}-title-${uniqueId}`;
+  }
+
+  /** @private */
+  __createContainer(slot) {
+    const container = document.createElement('div');
+    container.setAttribute('slot', slot);
+    return container;
+  }
+
+  /** @private */
+  __clearContainer(container) {
+    container.innerHTML = '';
+    // Whenever a Lit-based renderer is used, it assigns a Lit part to the node it was rendered into.
+    // When clearing the rendered content, this part needs to be manually disposed of.
+    // Otherwise, using a Lit-based renderer on the same node will throw an exception or render nothing afterward.
+    delete container._$litPart$;
+  }
+
+  /** @private */
+  __initContainer(container, slot) {
+    if (container) {
+      // Reset existing container in case if a new renderer is set.
+      this.__clearContainer(container);
+    } else {
+      // Create the container, but wait to append it until requestContentUpdate is called.
+      container = this.__createContainer(slot);
+    }
+    return container;
+  }
+
+  /** @private */
+  _headerFooterRendererChange(headerRenderer, footerRenderer, opened) {
+    const headerRendererChanged = this.__oldHeaderRenderer !== headerRenderer;
+    this.__oldHeaderRenderer = headerRenderer;
+
+    const footerRendererChanged = this.__oldFooterRenderer !== footerRenderer;
+    this.__oldFooterRenderer = footerRenderer;
+
+    const openedChanged = this._oldOpenedFooterHeader !== opened;
+    this._oldOpenedFooterHeader = opened;
+
+    if (headerRendererChanged) {
+      if (headerRenderer) {
+        this.headerContainer = this.__initContainer(this.headerContainer, 'header-content');
+      } else if (this.headerContainer) {
+        this.headerContainer.remove();
+        this.headerContainer = null;
+      }
+    }
+
+    if (footerRendererChanged) {
+      if (footerRenderer) {
+        this.footerContainer = this.__initContainer(this.footerContainer, 'footer');
+      } else if (this.footerContainer) {
+        this.footerContainer.remove();
+        this.footerContainer = null;
+      }
+    }
+
+    if (
+      (headerRenderer && (headerRendererChanged || openedChanged)) ||
+      (footerRenderer && (footerRendererChanged || openedChanged))
+    ) {
+      if (opened) {
+        this.requestContentUpdate();
+      }
+    }
+
+    this.toggleAttribute('has-header', !!headerRenderer);
+    this.toggleAttribute('has-footer', !!footerRenderer);
+  }
+
+  /** @private */
+  _headerTitleChanged(headerTitle, opened) {
+    if (opened && (headerTitle || this._oldHeaderTitle)) {
+      this.requestContentUpdate();
+    }
+    this._oldHeaderTitle = headerTitle;
+    this.toggleAttribute('has-title', !!headerTitle);
+  }
+
+  /** @private */
+  _headerTitleRenderer() {
+    if (this.headerTitle) {
+      if (!this.headerTitleElement) {
+        this.headerTitleElement = document.createElement('span');
```

Comment:
I was checking some component libraries and  the ones I found use `<h2>` for the dialog title. Should we use them here, as well? @jouni @web-padawan 

Examples: 
- https://www.material.io/components/dialogs
- https://www.radix-ui.com/docs/primitives/components/dialog

Reply:
We have a similar issue for `vaadin-confirm-dialog` header level (currently hardcoded to `<h3>`) and here is a comment that suggests not to hardcode it but let the developer choose: https://github.com/vaadin/web-components/issues/467#issuecomment-764531930.

Reply:
Let's create a follow-up issue and discuss this separately to not block this PR.